src/operator/window.rs

```diff
@@ -0,0 +1,403 @@
+    ///   monotonically, i.e., two successive invocations of `window_func` at
+    ///   logical times `t` and `t+1` must return `start_time1` and
+    ///   `start_time2` such that `start_time2 >= start_time1`.
+    ///
+    /// # Output
+    ///
+    /// The output stream contains **changes** to the contents of the window: at
+    /// every clock cycle it retracts values that belonged to the window at
+    /// the previous cycle, but no longer do, and inserts new values added
+    /// to the window.  The latter include new values in the input stream
+    /// that belong to the `[start_time..end_time)` range and values from
+    /// earlier inputs that fall within the new range, but not the previous
+    /// range.
+    pub fn window<F>(&self, window_func: F) -> Stream<Circuit<P>, OrdZSet<B::Val, B::R>>
+    where
+        F: FnMut() -> (B::Key, B::Key) + 'static,
+    {
+        let trace = self.integrate_trace().delay_trace();
+        self.circuit()
+            .add_binary_operator(<Window<B, F>>::new(window_func), &trace, self)
+    }
+}
+
+pub struct Window<B, F>
+where
+    B: IndexedZSet,
+{
+    window_func: F,
+    window: Option<(B::Key, B::Key)>,
+    _phantom: PhantomData<B>,
+}
+
+impl<B, F> Window<B, F>
+where
+    B: IndexedZSet,
+{
+    pub fn new(window_func: F) -> Self {
+        Self {
+            window_func,
+            window: None,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<B, F> Operator for Window<B, F>
+where
+    B: IndexedZSet,
+    F: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Window")
+    }
+    fn clock_start(&mut self, _scope: Scope) {
+        self.window = None;
+    }
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        // Windows can currently only be used in top-level circuits.
+        // Do we have meaningful examples of using windows inside nested scopes?
+        panic!("'Window' operator used in fixedpoint iteration")
+    }
+}
+
+impl<B, F> BinaryOperator<Spine<Rc<B>>, B, OrdZSet<B::Val, B::R>> for Window<B, F>
+where
+    B: IndexedZSet,
+    B::Val: Ord + Clone,
+    B::Key: Ord,
+    B::R: NegByRef,
+    F: FnMut() -> (B::Key, B::Key) + 'static,
+{
+    fn eval(&mut self, trace: &Spine<Rc<B>>, batch: &B) -> OrdZSet<B::Val, B::R> {
+        //           ┌────────────────────────────────────────┐
+        //           │       previous window                  │
+        //           │                                        │             e1
+        // ──────────┴────────────┬───────────────────────────┴──────────────┬─────►
+        //          s0          s1│                          e0              │
+        //                        │         new window                       │
+        //                        └──────────────────────────────────────────┘
+        //  region 1: [s0 .. s1)
+        //  region 2: [s1 .. e0)
+        //  region 3: [e0 .. e1)
+
+        let (start1, end1) = (self.window_func)();
```

Comment:
I tried to keep things simple initially. This design supports most useful forms of windows, except things like watermarks where window bounds depend on data.  But seeing that I'm not getting any slack here, I'll generalize this so that the operator reads window bounds from a separate stream.  That separate stream can be computed based on watermark data embedded in the time series or a physical time source.